### ISSUE TITLE
Malformedjsonexception in getlinks on page

### ISSUE THREAD

**Issue by kevinlinxc** [fastily/jwiki#21]
I get many of the following error when calling getLinksOnPage many times. 
It's similar to the login error from another issue, but when I try changing 
`new Wiki("en.wikipedia.org");` to 
`new Wiki("en.wikipedia.org/w/api.php"); `,
 nothing changed.
Here's the code that triggers the exception: https://hastebin.com/bopagefewo.cs
The strange thing is that the code still works after a while, but a stack trace full of thousands of these is not desirable.

com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Use 
com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 1 column 12 path $
	at com.google.gson.JsonParser.parse(JsonParser.java:65)
	at com.google.gson.JsonParser.parse(JsonParser.java:45)
	at fastily.jwiki.core.WQuery.next(WQuery.java:303)
	at fastily.jwiki.core.MQuery.getContProp(MQuery.java:72)
	at fastily.jwiki.core.MQuery.getLinksOnPage(MQuery.java:289)
	at fastily.jwiki.core.Wiki.getLinksOnPage(Wiki.java:799)
	at cpen221.mp3.wikimediator.WikiMediator.lambda$getConnectedPages$0(WikiMediator.java:175)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: com.google.gson.stream.MalformedJsonException: Use JsonReader.setLenient(true) to accept malformed JSON at line 1 column 12 path $
	at com.google.gson.stream.JsonReader.syntaxError(JsonReader.java:1568)
	at com.google.gson.stream.JsonReader.checkLenient(JsonReader.java:1409)
	at com.google.gson.stream.JsonReader.doPeek(JsonReader.java:542)
	at com.google.gson.stream.JsonReader.peek(JsonReader.java:425)
	at com.google.gson.JsonParser.parse(JsonParser.java:60)
	... 11 more

**Comment by fastily** [comment]
Hi @kevinlinxc, thanks for the bug report!

Could you provide a more specific example causing this error?

For example:
```java
import fastily.jwiki.core.Wiki;

public class Test
{
	public static void main(String[] args)
	{
		Wiki wiki = new Wiki("en.wikipedia.org");
		System.out.println(wiki.getLinksOnPage("Foobar"));
	}
}
```
works just fine for me.

Also, are you using the latest version of jwiki (1.7.0)?

**Comment by kevinlinxc** [comment]
Thanks for the quick reply. I am using 1.7.0, and getLinksOnPage work for me. This problem only happens when there are lots and lots of requests. which happens in my getConnecteddPages requests.
Here's a test that makes these errors:
@Test
    public void getConnectedPages1() {
            WikiMediator wm = new WikiMediator();
            List<String> foundList = wm.getConnectedPages("SpaceX", 3);
            assertTrue(foundList.contains("Boring Test Tunnel"));
    }
Constructor for WikiMediator:
public WikiMediator() {
        thisWiki = new Wiki("en.wikipedia.org/w/api.php");
        thisWiki.enableLogging(false);
        thisWiki.conf.debug = true;
}
Get connected pages is in the original hastebin.

**Comment by fastily** [comment]
Ah, that makes sense.  I'm guessing you're getting rate-limited by the Wikipedia API (on the server side).

If you have to access a ton of pages (in which case you should not use the API), Wikipedia provides database dumps for this exact reason, relevant links: [1](https://en.wikipedia.org/wiki/Wikipedia:Database_download), [2](https://dumps.wikimedia.org/backup-index.html).

If you must use the API, I recommend rate-limiting your application and keeping the number of threads to a minimum.

**Comment by kevinlinxc** [comment]
I must use the API for this, and so limiting to 2 threads is my only option but it takes forever as one might expect. Is there a way to get dumps from the API? The API allows for using | (pipe) to search for more things at once, so can I use that somehow with Jwiki?

**Comment by fastily** [comment]
> Is there a way to get dumps from the API?

No, the API is only for live data.

> The API allows for using | (pipe) to search for more things at once, so can I use that somehow with Jwiki?

You can try using the `getLinksOnPage` in the `MQuery`, but it is also subject to server-side rate limiting.

**Comment by fastily** [comment]
Procedurally closing this per inactivity and because it's not something I have any control over.

For future readers: 
**tl;dr** - please don't use the API for mass queries, the WMF _will_ rate-limit you.

Alternatives:
* [quarry](https://quarry.wmflabs.org) - WMF provided service that allows you to run SQL queries against the database replicas.
* [WMF DB Dumps](https://dumps.wikimedia.org/backup-index.html) - complete snapshots of all the data on just about every WMF project
